proof_of_concept/serialization.py

```diff
@@ -0,0 +1,290 @@
+"""(De)Serializes objects of various kinds to JSON."""
```

Comment:
Even if the (de)serialization code needs to be custom for each class, I would still prefer having a `from_json()` and `to_json()` method on the class itself.

Reply:
Introspection makes the code more difficult to understand though, I consider that a bit of an advanced feature. And so far all our classes have something funny about them that would complicated this. I started implementing to_json() and from_json(), but got stuck on the replica updates (which are a generic class) because typings/mypy isn't quite advanced enough. In C++ this would work beautifully, but well...

So I dropped that after spending way too much time on it (if I can find some time, I'll go write a feature proposal for typings), and instead decided to keep the aspect oriented design, but split up validation and serialization. That makes things a bit cleaner and means we're now doing the same thing with less code.

Reply:
I also tried to use YAtiML (with new JSON-functionality) but that didn't work because ruamel.yaml and cryptography's RSAPublicKey are fundamentally incompatible, and then I implemented a mini-serialiser to do the same thing using introspection, but it ended up being more lines of code and using more advanced Python features because things are quite irregular (Signature base class, RSAPublicKey, timestamp, class hierarchies with disambiguation fields, there are very few straightforward struct-type classes, so you end up with so many exceptions that you may as well write it by hand).

Reply:
I'm not using data classes (which are Python 3.7+ anyway and I like to have my code work on all supported Python versions), and jsonpickle explicitly says it's a security hole (although if you validate first it may be okay, but I don't feel for digging into it).

I'm going to stick with the aspect oriented design, putting all the serialization stuff in a separate module. I'm doing the same thing for signatures on objects (see signable.py), and it just seems like the best way of keeping the complexity under control.

Reply:
Okay, cleaned things up and it's looking nice now. Maybe a bit repetitive at first glance, but this is really the best solution, and a bit of repetition is also fine. DRY is not an iron law.